sdks/python/apache_beam/typehints/native_type_compatibility_test.py

```diff
@@ -103,6 +98,64 @@ def test_convert_to_beam_types(self):
         native_type_compatibility.convert_to_beam_types(typing_types),
         beam_types)
 
+  def test_is_sub_class(self):
+    self.assertTrue(native_type_compatibility._safe_issubclass(
+        parent=typing.Dict,
+        derived=typing.Dict[bytes, int]))
+    self.assertFalse(native_type_compatibility._safe_issubclass(
+        parent=typing.List,
+        derived=typing.Dict[bytes, int]))
+    self.assertTrue(native_type_compatibility._safe_issubclass(
+        parent=typing.List,
+        derived=typing.List[bytes]))
+    self.assertFalse(native_type_compatibility._safe_issubclass(
+        parent=typing.List,
+        derived=typing.Dict[bytes, int]))
+    self.assertTrue(native_type_compatibility._safe_issubclass(
+        parent=typing.Set,
+        derived=typing.Set[int]))
+    self.assertFalse(native_type_compatibility._safe_issubclass(
+        parent=typing.List,
+        derived=typing.Set[float]))
+    self.assertTrue(native_type_compatibility._safe_issubclass(
+        parent=typing.Tuple,
+        derived=typing.Tuple[int]))
+    self.assertFalse(native_type_compatibility._safe_issubclass(
+        parent=typing.List,
+        derived=typing.Tuple[bytes]))
+
+  @unittest.skipIf(sys.version_info >= (2, 7, 0),
```

Comment:
Why is this a correct behavior for this function to return different results on Py2 and Py3?

Reply:
To be honest I don't know if it is an advantage or not. I realised there was a difference and made the test to point it out. Maybe I also should have pointed it out more clearly in the PR as well. Based upon your comment I assume that we want to keep the Py2 behaviour and will investigate further if I can achieve that with minimal changes.

Reply:
I looked a little closer into the code of native_type_compatibility module.

The high-level goal of this change is to make sure that the correspondence between `typing` types and `apache_beam.typehints` types, defined by `convert_to_beam_type`, is established in the same way across all versions of Python. 

`test_convert_to_beam_type` should help us ensure this. As long as this test has sufficient coverage, we don't need to test all the helpers.  

`_is_subclass` is a helper function, and it's parent argument is one of typing.{Dict,List,Set}, so we can remove `_is_sublcass_order` test cases, which pass subscripted types, e.g. List[int].

There are several interesting differences in how following expressions are evaluated  different Py3.6 / Py3.7 versions, and there may be more differences in other Python versions.

```
type(typing.Any)
type(typing.Dict)
type(typing.List)
type(typing.Set)
type(typing.Tuple)
type(typing.Union)

type(typing.Any)
type(typing.Dict[int])
type(typing.List[int])
type(typing.Set[int])
type(typing.Tuple[int])
type(typing.Union[int,str])

typing.Dict[int].__origin__
typing.List[int].__origin__
typing.Set[int].__origin__
typing.Tuple[int].__origin__
typing.Union[int,str].__origin__

typing.Any.__origin__
typing.Dict.__origin__
typing.List.__origin__
typing.Set.__origin__
typing.Tuple.__origin__
typing.Union.__origin__
```
Taking these differences into account, we need to make sure that whatever comparisons we are doing in `_match_same_type` and `_is_subclass` make sense, and we not just making a change that makes a test pass. A reader unfamiliar with this code should also be able to understand the details of implementation. 

Looking at the evaluation of expressions above, I think we will be better off if we replace `_match_same_type` into separate `_match_union` and `_match_any` helpers.

Reply:
Yes, But what comparison makes sense. Would be like the comparison to go both ways or one way? I will remove the helper tests and split it.

Reply:
> But what comparison makes sense
It would help if you can make a recommendation, and explain why particular comparisons need to be made and why they make sense. It is sometimes easy to make a change to silence the test, but it should also be clear why a particular change is the right thing to do.